### ISSUE TITLE
[STORM-2449] Iterator of Redis State may return same key multiple time, with different values

### ISSUE THREAD

**Issue by jira-importer** [apache/storm#6231]
<p>Redis state iterator iterates pending prepare -> pending commit -> external storage (state) sequentially. While iterating, more of them are subject to change, so it can provide inconsistent result.</p>

<p>While we can't provide consistent result (since states are changing continuously), at least iterator needs to provide same key only once.</p>

---
<details><summary><i>Originally reported by <a title="Jungtaek Lim" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kabhwan">kabhwan</a>, imported from: <a href="https://issues.apache.org/jira/browse/STORM-2449" target="_blank">Iterator of Redis State may return same key multiple time, with different values</a></i></summary>
<i><ul>
<li><b>assignee</b>: <a title="Jungtaek Lim" href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kabhwan">kabhwan</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Major
<li><b>resolution</b>: Fixed
<li><b>resolved</b>: 2017-06-20T13:10:12+01:00
<li><b>imported</b>: 2025-01-24
</ul></i>
</details>

**Comment by jira-importer** [comment]
<i><a href="https://issues.apache.org/jira/secure/ViewProfile.jspa?name=kabhwan">kabhwan</a>:</i>
<p>Merged into master and 1.x branch.</p>
